lxd/storage/drivers/driver_cephfs.go

```diff
@@ -129,9 +129,67 @@ func (d *cephfs) Create() error {
 		fsPath = fields[1]
 	}
 
-	// Check that the filesystem exists.
-	if !d.fsExists(d.config["cephfs.cluster_name"], d.config["cephfs.user.name"], fsName) {
-		return fmt.Errorf("The requested '%v' CEPHFS doesn't exist", fsName)
+	// If the filesystem already exists, disallow keys associated to creating the filesystem.
+	fsExists := d.fsExists(d.config["cephfs.cluster_name"], d.config["cephfs.user.name"], fsName)
+	if fsExists {
+		for _, key := range []string{"create_missing", "osd_pg_num", "meta_pool", "data_pool"} {
+			cephfsSourceKey := fmt.Sprintf("cephfs.%s", key)
+			if d.config[cephfsSourceKey] != "" {
+				return fmt.Errorf("Invalid config key %q: CephFS filesystem already exists", cephfsSourceKey)
+			}
+		}
+	}
+
+	if !fsExists {
+		createMissing := shared.IsTrue(d.config["cephfs.create_missing"])
+		if !createMissing {
+			return fmt.Errorf("The requested '%v' CEPHFS doesn't exist", fsName)
+		}
+
+		// Default to 32 for the pg_num if otherwise unspecified.
+		pgNum := d.config["cephfs.osd_pg_num"]
+		if pgNum == "" {
+			d.config["cephfs.osd_pg_num"] = "32"
+		}
+
+		// Create the meta and data pools if necessary.
+		for _, key := range []string{"cephfs.meta_pool", "cephfs.data_pool"} {
+			pool := d.config[key]
+
+			if pool == "" {
+				return fmt.Errorf("Missing required key %q for creating cephfs osd pool", key)
+			}
+
+			if !d.osdPoolExists(d.config["cephfs.cluster_name"], d.config["cephfs.user.name"], pool) {
+				// Create new osd pool.
+				_, err := shared.TryRunCommand("ceph",
```

Comment:
I think I'd be tempted not to use TryRunCommand unless we know the reason why it fails intermittently. 
Ultimately I think it'd be better to remove TryRunCommand entirely eventually and instead add per-call site retries with comments as to why we are doing it, possibly with per-iteration error inspection, rather than having the blunt instrument that is this function.

Reply:
We should also add use of a reverter here, because if we create pool(s) and then fail later they will be left behind.

Reply:
This is taken directly from OSD pool creation on the `ceph` driver. So maybe we need to investigate why we're doing it that way already first.

Reply:
Yes, if we dont know why then take it out, is my view.

Reply:
As for the revert, I opted to not use one since the storage pool itself will still persist if there's an error, and with `create_missing=true` then we can just re-use what has already been created. Additionally we can't clean up when removing the `cephfs` storage pool because we can't be sure if they existed prior to creating the storage pool itself, since `create_missing` will also accept pre-existing OSD pools.

Reply:
> Additionally we can't clean up when removing the `cephfs` storage pool because we can't be sure if they existed prior to creating the storage pool itself, since `create_missing` will also accept pre-existing OSD pools.

In that case I think its even more important to have a reverted at create time.